### ISSUE TITLE
OpenAI API integration

### ISSUE THREAD

**Issue by JbIPS** [emilienschultz/activetigger#172]
Part of #163.

This issue is dedicated to [Open AI integration](https://platform.openai.com/docs/api-reference) discussion.

Concerning the [authentication](https://platform.openai.com/docs/api-reference/authentication), we'll need the `OPENAI_API_KEY` of the user to do the request. I know you'd rather not store them in DB, but the alternative is to attach it to _every requests_ between the client and the server, so the server could send it to OpenAI.

IMO, it is safer to store it, maybe encrypted, in the DB, knowing that OpenAI API keys can be generated for a designated scope and/or project, so the consequences of a DB leak would be limited. 

What's your opinion @emilienschultz?

**Comment by emilienschultz** [comment]
I follow your lead on this point. If you think the risk is limited (in the current architecture), let's go for that. 

We add a specific table for keys / generate parameters ?

**Comment by JbIPS** [comment]
I have not examine the DB schema yet, but it would make sense to store it in the same place as the account passwords

**Comment by JbIPS** [comment]
I was thinking about a clearer way to use the Generate page. My first thought was "users need to add a GenAI model to their project". So, at first, the page should be empty with only a button to "add a GenAI model".

Then, once the button clicked, a form (similar to the actual one) should be displayed, asking for a name/label (human-readable), model API (Llama/OpenAI), model name (`llama3.1`, `chatgpt-4o-mini`), endpoint (if Llama) or credentials (if OpenAI).

Those data would be store in DB and linked to the project. Once a GenAI model is linked, the prompt panel is shown and the generation can begin.

Obviously, we would need to add a switcher if multiple models are set.

**Advantages:**
1. Multiple models available, to compare/explore
2. Models are on a per-project basis: you could use a paid GPT (shared with your lab team) on one and a local on your private project
3. It's stored so you don't need to set it every time.

⚠️ Once a model is added, we shouldn't show whole the API_KEY (potentially to other collaborators on the project).

What do you think of that?

**Comment by emilienschultz** [comment]
Hi @JbIPS !

I see the idea, and I think it would be great from an ergonomic standpoint. 

I like the idea to specify endpoint for projects, it would be great. And allow modularity in the future.

I don't know actually if it is better to do fine-grained integration (possibility to select models with specific provider) or to limit to enter endpoints/credentials/type of API (and potentially human information on what's behind). I guess the low hanging fruits would be personal endpoint + huggingface. But maybe you found already existing way to integrate endpoints/api ?

What do you think @eollion ?

**Comment by JbIPS** [comment]
I'm not sure I understood the last point correctly. HuggingFace would be one more choice in the API selection, endpoint would be empty and user should provide a model and his HF token so we could call
```python
from huggingface_hub import InferenceClient

client = InferenceClient(
    "cardiffnlp/twitter-roberta-base-sentiment-latest",
    token="hf_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
)

client.text_classification("Today is a great day")
```
[source](https://huggingface.co/docs/api-inference/getting-started)

But I feel like I missed something in your comment 😅

**Comment by emilienschultz** [comment]
Sorry for the delay. Oh, I didn't know we could launch an endpoint directly with the huggingface client. When I thought of huggingface endpoint, I was rather thinking of this https://huggingface.co/inference-endpoints/dedicated so it is manual and then you get an endpoint. Or am I mistaken of the potential of the client ?

**Comment by JbIPS** [comment]
Thank you, I forgot about dedicated endpoints. But it shouldn't be an issue since you can simply [give the endpoint when choosing the model](https://huggingface.co/docs/huggingface_hub/guides/inference#using-a-specific-url):
```python
from huggingface_hub import InferenceClient
client = InferenceClient(model="https://uu149rez6gw9ehej.eu-west-1.aws.endpoints.huggingface.cloud/deepfloyd-if")
client = InferenceClient()
client.text_to_image(..., model="https://uu149rez6gw9ehej.eu-west-1.aws.endpoints.huggingface.cloud/deepfloyd-if")
```

Or I you suggesting to create dedicated endpoints directly into ActiveTigger? This could be done, since there is an API for that, but I doubt it is on the scope of the project.

**Comment by emilienschultz** [comment]
No, you're right, I think the creation of endpoint should be separated from AT interface
